proposals/data-objects.md

```diff
@@ -1,82 +1,207 @@
+* **Authors**: Alexander Udalov, Roman Elizarov, Pavel Mikhailovskii, Marat Akhin
+* **Status**: Preview in 1.8.20, Release in 1.9.0
 * **Discussion and feedback**: [#317](https://github.com/Kotlin/KEEP/issues/317)
 
+# Summary
 
-## Introduction
+This KEEP introduces ***data objects*** which fix the inconsistencies in the current Kotlin design related to how one works with immutable data and algebraic data types (ADTs) via data classes, and how to avoid the boilerplate of implementing default `toString` for objects. As one of the effects, it fixes the [KT-4107](https://youtrack.jetbrains.com/issue/KT-4107) feature request.
 
-When using algebraic data types, it’s annoying to declare singleton values because there’s no way to get automatic toString similar to data classes. E.g. on JVM:
-```kotlin
-sealed class MyList<T> {
-    data class Cons<T>(val value: T) : MyList<T>()
-    object Nil : MyList<Nothing>()
+# Motivation
+
+## Current state
+
+Currently, when working with regular data entities which follow the standard rules, one can use [data classes](https://kotlinlang.org/docs/data-classes.html) to avoid the need to manually create a number of utility functions with known standard behavior, as they can be automatically generated by the compiler. For a data class with data properties `{pi}`, we have the following functions [generated](https://kotlinlang.org/spec/declarations.html#data-class-declaration).
+
+* `equals`/`hashCode` which follow the structural equality rules and consider the pairwise equality of `{pi}`
+* `toString` containing the data class name together with its data properties’ string representations
+* `copy` to support immutability by facilitating easy data class copying while changing one or more data property
+* `componentN` for destructuring declaration
+
+This allows one of the ways you use data classes in Kotlin: as data holders with a number of convenience features. They are most useful for representing immutable value-like data, but it is not a hard constraint. If you need, you can use data classes with mutable data, or even mark your class as a data class for the single purpose of getting a generated `toString` implementation.
+
+The second way one may use data classes is an extension of their “immutable data” nature. Together with the [sealed types](https://kotlinlang.org/docs/sealed-classes.html), data classes allow to describe ADTs in a convenient way, where the sum type is represented as a sealed class/interface, and the product type is represented as a data class.
+
+```
+// An ADT representing simple arithmetic expressions
+sealed interface Expr
+
+data class Add(val lhv: Expr, val rhv: Expr) : Expr
+data class Sub(val lhv: Expr, val rhv: Expr) : Expr
+data class Mul(val lhv: Expr, val rhv: Expr) : Expr
+data class Div(val lhv: Expr, val rhv: Expr) : Expr
+data class Const(val value: Int) : Expr
+
+fun eval(e: Expr): Int = when (e) {
+    is Add -> *eval*(e.lhv) + *eval*(e.rhv)
```

Comment:
Emphasis inside code is not rendered. You'd better replace `*eval*` with `eval`.